src/mem/alloc.h

```diff
@@ -62,20 +62,112 @@ namespace snmalloc
     Pagemap<SUPERSLAB_BITS, uint8_t, 0>>;
 
   HEADER_GLOBAL SuperslabPagemap global_pagemap;
+
+  /**
+   * Mixin used by `SuperslabMap` to directly access the pagemap via a global
+   * variable.  This should be used from within the library or program that
+   * owns the pagemap.
+   */
+  struct GlobalPagemap
+  {
+    /**
+     * Returns the pagemap.
+     */
+    SuperslabPagemap& pagemap()
+    {
+      return global_pagemap;
+    }
+  };
+
+  /**
+   * Mixin used by `SuperslabMap` to access a pagemap provided by the user.
+   * This should be used by code that needs to access the canonical pagemap via
+   * some custom mechanism.
+   */
+  class CustomPagemap
+  {
+    /**
+     * Reference to the pagemap that we're using, allowing it to not be the
+     * default pagemap.
+     */
+    SuperslabPagemap& custom_pagemap;
+
+  public:
+    /**
+     * Constructor.  Takes the pagemap to use as an argument.  This constructor
+     * will be inherited by `SuperslabMap` if this mixin is used and so should
+     * be called as a constructor there, rather than directly.
+     */
+    CustomPagemap(SuperslabPagemap& pm) : custom_pagemap(pm) {}
+
+    /**
+     * Returns the provided pagemap.
+     */
+    SuperslabPagemap& pagemap()
+    {
+      return custom_pagemap;
+    }
+  };
+
+  /**
+   * Optionally exported function that accesses the global pagemap provided by
+   * a shared library.
+   */
+  extern "C" void* snmalloc_pagemap_global_get(snmalloc::PagemapConfig const**);
+
+  /**
+   * Mixin used by `SuperslabMap` to access the global pagemap via a
+   * type-checked C interface.  This should be used when another library (e.g.
+   * your C standard library) uses snmalloc and you wish to use a different
+   * configuration in your program or library, but wish to share a pagemap so
+   * that either version can deallocate memory.
+   */
+  class ExternalGlobalPagemap
+  {
+    /**
+     * A pointer to the pagemap.
+     */
+    SuperslabPagemap* external_pagemap;
+
+  public:
+    /**
+     * Construtor.  Accesses the pagemap via the C ABI accessor and casts it to
+     * the expected type, failing in cases of ABI mismatch.
+     */
+    ExternalGlobalPagemap()
+    {
+      const snmalloc::PagemapConfig* c;
+      external_pagemap =
+        SuperslabPagemap::cast_to_pagemap(snmalloc_pagemap_global_get(&c), c);
```

Comment:
This can't be used, if `snmalloc_pagemap_global_get` ignores the config.  Should we make cast able to take a null config and just succeed?